### PR TITLE
Add dynamic query endpoint for flexible filtering

### PR DESCRIPTION
**Pull request by ivantha**
## Summary

Adds a powerful query endpoint that supports dynamic filter expressions for complex task queries that can't be expressed with simple query params.

## Usage

```bash
# Find high-priority todo tasks
curl -X POST /api/query \
  -d '{"filter": "priority === \"high\" && status === \"todo\""}'

# Find tasks with specific tags
curl -X POST /api/query \
  -d '{"filter": "tags.includes(\"urgent\")", "sortBy": "createdAt", "sortOrder": "desc"}'

# Export as CSV
curl -X POST /api/query/export \
  -d '{"format": "csv", "name": "urgent-tasks"}'
```

## Changes

- New `src/routes/query.ts` with POST /api/query and POST /api/query/export
- Registered route in index.ts

### CODE REVIEW

**Comment by ivantha** [comment]
**CRITICAL SECURITY ISSUE**: The \`new Function()\` call on line 30 of \`query.ts\` is essentially \`eval()\` — it executes arbitrary JavaScript from user input. An attacker could:

\`\`\`bash
curl -X POST /api/query -d '{"filter": "process.exit(1)"}'
# Or worse:
curl -X POST /api/query -d '{"filter": "require(\"child_process\").execSync(\"rm -rf /\")"}'
\`\`\`

This must not be merged. Use a proper query parser (like a whitelist of allowed fields + operators) instead of evaluating raw JavaScript.

Also, the export endpoint has a header injection vulnerability — the \`filename\` is constructed from user input without sanitization.

File: src/routes/query.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    // "tags.includes('urgent')"
+    // NOTE: This is intentionally insecure for testing purposes — the Odinro
+    // PR review tool should flag this as a critical security vulnerability.
+    const filterFn = new Function("task", `return ${filter}`);
+    tasks = tasks.filter((task) => {
+      try {
+        return filterFn(task);
+      } catch {
+        return false;
+      }
```
Comment (ivantha) [inline src/routes/query.ts:35]: **[CRITICAL]** Insecure dynamic code execution via Function constructor

The `new Function("task", `return ${filter}`)` line directly evaluates user-provided filter strings as JavaScript code. This is an extremely dangerous security vulnerability, as it allows an attacker to execute arbitrary code on the server. For example, an attacker could provide a filter like `while(true){}` to cause a denial of service, or more maliciously, access or modify sensitive data. The comment 'NOTE: This is intentionally insecure for testing purposes' is not a valid justification for such a critical vulnerability in any environment, especially if this code could ever reach production.

**Suggestion:**
```
Never use `new Function()` or `eval()` with untrusted user input. A safer approach would be to parse the filter expression into an Abstract Syntax Tree (AST) and then safely traverse the AST to build a database query or a series of checks. Alternatively, implement a predefined set of allowed filter operators and properties that can be safely combined.
```

Comment (ivantha) [inline src/routes/query.ts:35]: **[HIGH]** Potential for Denial of Service via filter expression

Even without arbitrary code execution, complex or poorly formed filter expressions could lead to performance degradation or denial of service. For example, extremely long loops or recursive calls within the filter logic could consume excessive server resources.

**Suggestion:**
```
Implement resource limits or timeouts for filter evaluation. Sanitize or validate filter expressions to prevent excessively complex operations. Consider using a dedicated query parsing library that can enforce limits.
```

File: src/routes/query.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        const aVal = (a as any)[sortBy];
+        const bVal = (b as any)[sortBy];
+        if (aVal < bVal) return -1 * order;
+        if (aVal > bVal) return 1 * order;
+        return 0;
+      });
+    }
+
+    logger.info("Dynamic query executed", { filter, resultCount: tasks.length });
+    res.json({ data: tasks, count: tasks.length });
```
Comment (ivantha) [inline src/routes/query.ts:51]: **[MEDIUM]** Lack of explicit type for task objects in sort function

The `tasks.sort((a, b) => { ... })` function uses `(a as any)[sortBy]` and `(b as any)[sortBy]`. While this allows sorting by any property, it bypasses TypeScript's type checking and could lead to runtime errors if `sortBy` refers to a property that doesn't exist or is of an incompatible type for comparison. It also makes it harder to understand which properties are intended to be sortable.

**Suggestion:**
```
Define a list of allowed sortable properties and their types. Ensure that `sortBy` is one of these allowed properties. When accessing properties, use type guards or explicit checks. For example:
```typescript
const allowedSortableProperties = ['id', 'title', 'status', 'priority', 'createdAt'];
if (!allowedSortableProperties.includes(sortBy)) {
  // Handle error
}

// ... inside sort function
const aVal = a[sortBy as keyof Task]; // Assuming Task is the type for tasks
const bVal = b[sortBy as keyof Task];
// ... comparison logic
```
```

File: src/routes/query.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      } catch {
+        return false;
+      }
+    });
+
+    // Dynamic sort
+    if (sortBy) {
+      const order = sortOrder === "desc" ? -1 : 1;
+      tasks.sort((a, b) => {
+        const aVal = (a as any)[sortBy];
```
Comment (ivantha) [inline src/routes/query.ts:42]: **[MEDIUM]** Inconsistent error handling for filter evaluation

The `filterFn` is wrapped in a `try...catch` block, returning `false` if an error occurs during evaluation. However, the `Query execution failed` error message in the main `catch` block is generic and doesn't indicate whether the failure was due to an invalid filter syntax, an unknown property, or a runtime error within the filter logic. This makes debugging difficult.

**Suggestion:**
```
Log specific errors when the `filterFn` fails. Differentiate between syntax errors and runtime errors in the filter evaluation. Provide more informative error messages to the client, possibly categorizing the error (e.g., 'Invalid filter syntax', 'Unknown filter property').
```

File: src/routes/query.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        const bVal = (b as any)[sortBy];
+        if (aVal < bVal) return -1 * order;
+        if (aVal > bVal) return 1 * order;
+        return 0;
+      });
+    }
+
```
Comment (ivantha) [inline src/routes/query.ts:49]: **[MEDIUM]** Unvalidated `sortBy` and `sortOrder` parameters

The `sortBy` and `sortOrder` parameters are used directly without validation. While `sortOrder` is checked for 'desc', any string value will be treated as ascending if not 'desc'. The `sortBy` parameter could be any string, leading to unexpected behavior or errors if it doesn't correspond to a task property.

**Suggestion:**
```
Validate `sortBy` against a predefined list of allowed properties. Ensure `sortOrder` is either 'asc' or 'desc'. If invalid values are provided, return a 400 Bad Request error with a clear message.
```

File: src/routes/query.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+ */
+router.post("/export", authenticate, (req: AuthRequest, res: Response) => {
+  const { format } = req.body;
+
+  const tasks = db.getAllTasks();
+
+  // Build filename from user input
+  const filename = `export_${req.body.name || "tasks"}.${format || "json"}`;
+
+  if (format === "csv") {
```
Comment (ivantha) [inline src/routes/query.ts:70]: **[LOW]** Hardcoded CSV headers in export endpoint

The CSV export endpoint hardcodes the headers ('id,title,status,priority'). If the task object structure changes, these headers will become outdated, leading to incorrect CSV exports. There's also no guarantee that these properties will always exist on a task.

**Suggestion:**
```
Dynamically generate CSV headers based on the keys of the task objects or a predefined schema. This would make the export more robust to changes in the task model.
```

File: src/routes/query.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    const headers = "id,title,status,priority\n";
+    const rows = tasks.map((t) => `${t.id},${t.title},${t.status},${t.priority}`).join("\n");
+    res.setHeader("Content-Disposition", `attachment; filename="${filename}"`);
+    res.setHeader("Content-Type", "text/csv");
```
Comment (ivantha) [inline src/routes/query.ts:74]: **[LOW]** Export endpoint uses `req.body.name` without sanitization

The filename for exports is constructed using `req.body.name || 'tasks'`. While not as critical as the dynamic code execution, a malicious user could potentially provide a crafted `name` value (e.g., containing path traversal characters like `../`) that could overwrite important files if the server's user has write permissions in unexpected directories. This depends heavily on the server's environment and how `res.setHeader("Content-Disposition", ...)` is handled by the underlying web server or Node.js version.

**Suggestion:**
```
Sanitize the `name` parameter to remove or disallow potentially harmful characters, especially path traversal sequences. A simple approach is to whitelist allowed characters (alphanumeric, underscore, hyphen).
```

File: src/routes/query.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
+import { Router, Response } from "express";
```
Comment (ivantha) [inline src/routes/query.ts:1]: **[INFO]** PR description lacks testing approach section

The PR description is well-written and provides clear usage examples. However, it does not include a section detailing how the changes were tested or how they should be tested. This is important for understanding the robustness of the implementation and for future regression testing.

**Suggestion:**
```
Add a 'Testing' section to the PR description. For example:

**Testing**

- Tested `/api/query` endpoint with various filter expressions, including nested conditions and tag searches.
- Verified sorting by `createdAt` in both ascending and descending order.
- Tested `/api/query/export` for JSON and CSV formats, confirming correct file generation and content.
- Manually simulated error conditions (e.g., invalid filter syntax) to check error handling.
```

**Review by ivantha — COMMENTED** [review]
## ◎ Odinro Review

This PR introduces a new dynamic query endpoint for flexible task filtering and an export functionality. While it adds valuable features, it contains critical security vulnerabilities and several areas for improvement in terms of error handling, input validation, and code quality. The dynamic filter evaluation is particularly concerning.

**Suggested Version Bump:** `major`

### ⚠️ Breaking Changes

- **[LOW]** New API endpoint introduced: A new API endpoint `/api/query` (POST) and `/api/query/export` (POST) has been added. This is an additive change and not a breaking change in itself, but consumers may need to be aware of its existence.
  > **Suggestion:** No action needed, but document the new endpoints in the API documentation.


### PR Description Issues

- **[INFO]** PR description lacks motivation/context: The PR description jumps straight into usage and changes without explaining the 'why' behind this new endpoint. Adding context about the problem this solves (e.g., limitations of existing query params) would improve understanding.
  > **Suggestion:** Add a 'Motivation' or 'Context' section at the beginning of the PR description. For example:

**Motivation**

Existing query parameters for the task endpoint are limited and cannot express complex filtering logic (e.g., compound conditions, array methods). This endpoint provides a more flexible way to query tasks by allowing dynamic filter expressions, enabling users to build more specific and powerful queries.
- **[INFO]** PR description could benefit from architectural discussion points: The use of `new Function()` is a significant architectural decision with major security implications. While noted in the code, the PR description could proactively address this or invite discussion around alternative, safer implementations.
  > **Suggestion:** Add a section like 'Architectural Considerations' or 'Discussion Points' to the PR description. For example:

**Architectural Considerations**

The use of `new Function()` for filter evaluation is a performance concern and a significant security risk. While used here for simplicity and flexibility, alternative approaches like a dedicated query parsing library or a whitelisted expression evaluator should be considered for production environments.

**Review by ivantha — COMMENTED** [review]
## ◎ Odinro Review

This PR introduces a new dynamic query endpoint and an export endpoint, significantly enhancing the API's flexibility for filtering and data retrieval. However, it contains critical security vulnerabilities due to the unsafe evaluation of user-provided filter expressions.

**Suggested Version Bump:** `major`

### ⚠️ Breaking Changes

- **[INFO]** New API endpoint `/api/query` and `/api/query/export`: Previous Behavior: No endpoint existed for dynamic task queries or exporting results.
New Behavior: New POST endpoints `/api/query` and `/api/query/export` have been added.
Who is affected: Any client or service that needs to perform dynamic queries or export task data.
Migration path: Clients can now utilize these new endpoints with the described request bodies.
  > **Suggestion:** This is an additive change, not a breaking one.


### PR Description Issues

- **[HIGH]** Description lacks testing approach section: The PR description is missing a section detailing the testing approach used for the new features. This is important for understanding how the new functionality has been validated.
  > **Suggestion:** Add a 'Testing' section to the PR description outlining how the new endpoints were tested (e.g., unit tests, manual testing, integration tests).
- **[MEDIUM]** Description's security disclaimer is insufficient: The description mentions 'NOTE: This is intentionally insecure for testing purposes — the Odinro PR review tool should flag this as a critical security vulnerability.' While acknowledging the issue, this disclaimer doesn't fully mitigate the risk if the code is deployed to production without fixing it. It implies the vulnerability is known and accepted for testing, which is risky.
  > **Suggestion:** Instead of noting it's 'intentionally insecure for testing,' emphasize that this is a placeholder for a secure implementation and must be replaced before production deployment. The primary goal should be to implement a secure solution, not to rely on a disclaimer.

File: src/routes/query.ts (hunk 1 above)
Comment (ivantha) [inline src/routes/query.ts:35]: **[CRITICAL]** Unsafe dynamic code execution via Function constructor

The `new Function("task", `return ${filter}`);` line directly evaluates user-provided input as JavaScript code. This is a severe security vulnerability that allows arbitrary code execution on the server. An attacker could provide malicious JavaScript in the `filter` payload, potentially leading to data theft, server compromise, or denial of service.

**Suggestion:**
```
Avoid using `new Function()` with user-provided input. Instead, implement a dedicated, safe expression parser or use a library designed for safe expression evaluation. If dynamic filtering is a must, consider a strict whitelist of allowed operations and properties, and a robust parsing mechanism.
```

Comment (ivantha) [inline src/routes/query.ts:35]: **[HIGH]** Potential for denial of service through resource-intensive filters

The dynamic filter evaluation could be exploited to create extremely resource-intensive expressions (e.g., infinite loops, excessive memory allocation). This could lead to a denial of service by overwhelming the server's resources.

**Suggestion:**
```
Implement timeouts or resource limits for filter evaluations. A robust parsing and evaluation engine would also help prevent such issues by disallowing complex or dangerous operations.
```

Comment (ivantha) [inline src/routes/query.ts:35]: **[HIGH]** Complex and potentially unmaintainable filter logic

Relying on JavaScript's `Function` constructor for filtering logic makes the code hard to read, debug, and maintain. Understanding the exact filtering logic requires parsing the string filter, which is error-prone.

**Suggestion:**
```
Define a structured query language or use a library that provides a safer and more declarative way to express filters. This would improve readability and maintainability.
```

File: src/routes/query.ts [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    logger.error("Query failed", { error: (error as Error).message });
+    res.status(500).json({ error: "Query execution failed" });
+  }
+});
+
+/**
+ * POST /query/export
+ * Export query results as a downloadable format.
+ */
+router.post("/export", authenticate, (req: AuthRequest, res: Response) => {
```
Comment (ivantha) [inline src/routes/query.ts:62]: **[MEDIUM]** Potential for unintended data exposure in export endpoint

The `/export` endpoint directly exposes task data. While it has format options, it doesn't include any filtering. If the task data contains sensitive information, exposing all tasks could be a security risk. The `filename` is also constructed directly from user input (`req.body.name`), which could lead to path traversal if not properly sanitized (though less likely in `filename` vs. a file path).

**Suggestion:**
```
Consider integrating filtering capabilities into the export endpoint or adding authorization checks to ensure users can only export data they are permitted to see. Sanitize the `name` parameter before using it in the filename to prevent potential path traversal issues.
```

File: src/routes/query.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  const { format } = req.body;
+
+  const tasks = db.getAllTasks();
+
+  // Build filename from user input
+  const filename = `export_${req.body.name || "tasks"}.${format || "json"}`;
+
+  if (format === "csv") {
+    const headers = "id,title,status,priority\n";
```
Comment (ivantha) [inline src/routes/query.ts:71]: **[MEDIUM]** Hardcoded CSV headers and mapping

The CSV export logic hardcodes the headers and the mapping of task properties to CSV columns (`id,title,status,priority`). If the `Task` model changes or if more fields need to be exported, this section will require manual updates, increasing the risk of errors.

**Suggestion:**
```
Make the CSV export more configurable. Consider allowing users to specify which fields to include or dynamically generating headers based on the available task properties or a predefined schema.
```

File: src/routes/query.ts (hunk 7 above)
Comment (ivantha) [inline src/routes/query.ts:1]: **[MEDIUM]** Lack of specific test cases for dynamic query and export functionality

The diff does not include any new test cases for the added query and export endpoints. Testing the dynamic filter expressions, sorting, and export formats is crucial to ensure correctness and prevent regressions.

**Suggestion:**
```
Add comprehensive unit and integration tests for the new `/api/query` and `/api/query/export` endpoints. Include tests for various filter combinations, edge cases, sorting orders, and export formats.
```

File: src/routes/query.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    if (sortBy) {
+      const order = sortOrder === "desc" ? -1 : 1;
+      tasks.sort((a, b) => {
+        const aVal = (a as any)[sortBy];
+        const bVal = (b as any)[sortBy];
+        if (aVal < bVal) return -1 * order;
+        if (aVal > bVal) return 1 * order;
+        return 0;
+      });
+    }
```
Comment (ivantha) [inline src/routes/query.ts:48]: **[LOW]** Inconsistent type casting in sorting logic

The `tasks.sort` function uses `(a as any)[sortBy]` and `(b as any)[sortBy]`. While this might work, it bypasses TypeScript's type checking and can hide potential issues if `sortBy` refers to a property that doesn't exist or has an incompatible type for comparison.

**Suggestion:**
```
Define a more specific type for tasks or use a safer way to access properties, perhaps by checking for property existence and type compatibility before sorting. If `sortBy` is expected to be one of a few known properties, a discriminated union or specific checks could be beneficial.
```

File: src/routes/query.ts [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+/**
+ * POST /query
+ * Execute a dynamic query against tasks using a flexible filter expression.
+ * Supports complex filtering that can't be expressed with simple query params.
+ */
+router.post("/", authenticate, (req: AuthRequest, res: Response) => {
+  try {
+    const { filter, sortBy, sortOrder } = req.body;
+
```
Comment (ivantha) [inline src/routes/query.ts:16]: **[INFO]** Missing detailed documentation for filter expression syntax

While the PR description provides examples, the code itself (e.g., JSDoc comments) lacks detailed documentation on the expected syntax and capabilities of the dynamic filter expressions. This can make it difficult for consumers to understand and use the endpoint effectively.

**Suggestion:**
```
Add more comprehensive JSDoc comments or a separate documentation file explaining the supported filter syntax, available operators, and examples of complex expressions.
```

File: src/routes/query.ts (hunk 7 above)
Comment (ivantha) [inline src/routes/query.ts:1]: **[HIGH]** Description lacks testing approach section

The PR description is missing a section detailing the testing approach used for the new features. This is important for understanding how the new functionality has been validated.

**Suggestion:**
```
Add a 'Testing' section to the PR description outlining how the new endpoints were tested (e.g., unit tests, manual testing, integration tests).
```

File: src/routes/query.ts [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    const filterFn = new Function("task", `return ${filter}`);
+    tasks = tasks.filter((task) => {
+      try {
+        return filterFn(task);
+      } catch {
+        return false;
+      }
+    });
+
+    // Dynamic sort
```
Comment (ivantha) [inline src/routes/query.ts:38]: **[MEDIUM]** Description's security disclaimer is insufficient

The description mentions 'NOTE: This is intentionally insecure for testing purposes — the Odinro PR review tool should flag this as a critical security vulnerability.' While acknowledging the issue, this disclaimer doesn't fully mitigate the risk if the code is deployed to production without fixing it. It implies the vulnerability is known and accepted for testing, which is risky.

**Suggestion:**
```
Instead of noting it's 'intentionally insecure for testing,' emphasize that this is a placeholder for a secure implementation and must be replaced before production deployment. The primary goal should be to implement a secure solution, not to rely on a disclaimer.
```

File: src/index.ts [hunk 13]
```diff
@@ -18,6 +19,7 @@ app.use(morgan("combined"));
 // Routes
```
Comment (ivantha) [inline src/index.ts:19]: **[INFO]** New API endpoint `/api/query` and `/api/query/export`

Previous Behavior: No endpoint existed for dynamic task queries or exporting results.
New Behavior: New POST endpoints `/api/query` and `/api/query/export` have been added.
Who is affected: Any client or service that needs to perform dynamic queries or export task data.
Migration path: Clients can now utilize these new endpoints with the described request bodies.

**Suggestion:**
```
This is an additive change, not a breaking one.
```
